### ISSUE TITLE
[AMD][kernel][Tier 5] Port sol-execbench Quant/030_nvfp4_fused_decoder_layer_pre_norm_attention to FlyDSL

### ISSUE THREAD

**Issue by sunxxuns** [amdpilot-org/FlyDSL#174]
# FlyDSL implementation: port sol-execbench's `030_nvfp4_fused_decoder_layer_pre_norm_attention` problem to FlyDSL

## Goal

Port the `030_nvfp4_fused_decoder_layer_pre_norm_attention` problem from `nvidia/SOL-ExecBench` (subset `Quant`) to FlyDSL as part of the sol-execbench → FlyDSL coverage sweep. This is one of 235 problems across 4 subsets (L1, L2, Quant, FlashInfer-Bench) being tracked for FlyDSL kernel coverage.

**Source:** `nvidia/Qwen3-14B-NVFP4` (per `definition.json` `hf_id`)

**Description:** FP4-quantized fused decoder layer pre-normalization and complete attention computation. Combines RMSNorm on input hidden states, Q/K/V projections with FP4 quantization, per-head RMSNorm on Q and K, RoPE positional embeddings, Grouped Query Attention with KV repeat, and output projection with FP4 quantization.

## Hardware / workload

```text
Hardware:  1 × MI300X (gfx942) + 1 × MI355X (gfx950)
Models:    (none — kernel benchmark, not LLM workload)
Subset:    Quant
Problem:   030_nvfp4_fused_decoder_layer_pre_norm_attention
Source:    nvidia/Qwen3-14B-NVFP4
```

**Const axes:**
  - `hidden_size` = 5120
  - `num_attention_heads` = 40
  - `num_key_value_heads` = 8
  - `head_dim` = 128

**Inputs:**
  - `hidden_states` [batch_size, seq_len, hidden_size] bfloat16 — Input hidden states
  - `cos` [batch_size, seq_len, head_dim] bfloat16 — Cosine for RoPE
  - `sin` [batch_size, seq_len, head_dim] bfloat16 — Sine for RoPE
  - `attention_mask` [batch_size, 1, seq_len, seq_len] bfloat16 — Attention mask
  - `input_layernorm_weight` [hidden_size] bfloat16 — Input layer norm weight
  - `q_weight_fp4` [q_out_features, hidden_size_half] float4_e2m1fn_x2 — Q projection FP4 weight
  - `q_scale` [q_out_features, hidden_scale_blocks] float8_e4m3fn — Q projection scale factors
  - `q_global_scale` [] float32 — Q projection global scale
  - `k_weight_fp4` [kv_out_features, hidden_size_half] float4_e2m1fn_x2 — K projection FP4 weight
  - `k_scale` [kv_out_features, hidden_scale_blocks] float8_e4m3fn — K projection scale factors
  - `k_global_scale` [] float32 — K projection global scale
  - `v_weight_fp4` [kv_out_features, hidden_size_half] float4_e2m1fn_x2 — V projection FP4 weight
  - `v_scale` [kv_out_features, hidden_scale_blocks] float8_e4m3fn — V projection scale factors
  - `v_global_scale` [] float32 — V projection global scale
  - `o_weight_fp4` [hidden_size, q_out_half] float4_e2m1fn_x2 — Output projection FP4 weight
  - `o_scale` [hidden_size, q_out_scale_blocks] float8_e4m3fn — Output projection scale factors
  - `o_global_scale` [] float32 — Output projection global scale
  - `q_norm_weight` [head_dim] bfloat16 — Per-head Q norm weight
  - `k_norm_weight` [head_dim] bfloat16 — Per-head K norm weight
  - `rms_norm_eps` [] float32 — RMSNorm epsilon

**Outputs:**
  - `attn_output` [batch_size, seq_len, hidden_size] bfloat16 — Attention output

**Workload samples (first 3):**
  - {"uuid": "a3bd7b42-402d-5dc8-90a2-cad4676e5f85", "axes": {"batch_size": 4, "seq_len": 256}, "inputs": {"hidden_states": {"type": "custom"}, "cos": {"type": "custom"}, "sin": {"type": "custom"}, "atten
  - {"uuid": "bd03fb1f-17fc-5d7e-a4cb-1029b374f1a4", "axes": {"batch_size": 64, "seq_len": 128}, "inputs": {"hidden_states": {"type": "custom"}, "cos": {"type": "custom"}, "sin": {"type": "custom"}, "atte
  - {"uuid": "56feee66-fa87-5e67-8542-1ad74e6c8524", "axes": {"batch_size": 2, "seq_len": 2048}, "inputs": {"hidden_states": {"type": "custom"}, "cos": {"type": "custom"}, "sin": {"type": "custom"}, "atte

```text
Baseline:  sol-execbench PyTorch reference (reference.py) on AMD with torch.compile off
Target:    max_abs_diff < 1e-2 vs PyTorch reference on all tested shapes
           kernel_ms < 0.9 × PyTorch eager baseline on ≥80% of workload shapes
```

## Scope

1. `kernels/sol_execbench/quant_030_nvfp4_fused_decoder_layer_pre_norm_attention.py` — FlyDSL kernel implementing the op described in `definition.json`, matching the PyTorch reference's numerics
2. `tests/kernels/test_sol_execbench_quant_030_nvfp4_fused_decoder_layer_pre_norm_attention.py` — parity harness vs PyTorch reference, testing multiple workload shapes
3. `tests/kernels/bench_sol_execbench_quant_030_nvfp4_fused_decoder_layer_pre_norm_attention.py` — reports `kernel_ms` per shape + comparison vs PyTorch eager baseline
4. `assessment.md` (≤300 words) — ranking which shapes FlyDSL wins/ties/loses vs PyTorch eager

## Acceptance

- Correctness: max_abs_diff < 1e-2 vs PyTorch reference on all tested workload shapes
- Performance: kernel_ms < 0.9 × PyTorch eager baseline on ≥80% of workload shapes
- Coverage: works on both gfx942 (MI300X) and gfx950 (MI355X)

## Out of scope

- Backward pass (unless the problem is itself a backward op)
- Weight gradients
- Optimizer step
- Cross-arch comparison vs NVIDIA H100 (PyTorch eager on AMD is the baseline)

## Reproduction

Clone `amdpilot-org/SOL-ExecBench`, download the dataset (`scripts/download_data.sh`), find the problem at `data/benchmark/Quant/030_nvfp4_fused_decoder_layer_pre_norm_attention/`, run the sol-execbench driver against a FlyDSL solution. Requires the unified aiter+triton+flydsl Docker image at `/home/xiasun/unified-rocm.Dockerfile` (not yet built).

## Definition of done

- [ ] Kernel implemented at `kernels/sol_execbench/quant_030_nvfp4_fused_decoder_layer_pre_norm_attention.py`
- [ ] Tests pass on MI300X (gfx942)
- [ ] Tests pass on MI355X (gfx950)
- [ ] Bench shows ≥80% win rate vs PyTorch eager
- [ ] `assessment.md` written

## ## Leaderboard baseline to beat

These are the current best public results on NVIDIA B200 (from the
[sol-execbench leaderboard](https://research.nvidia.com/benchmarks/sol-execbench/leaderboard))
for problem `030_nvfp4_fused_decoder_layer_pre_norm_attention` (kernel_id=206). The agent's
FlyDSL kernel on MI300X/MI355X should aim to beat the **Scoring Baseline**
(optimized PyTorch) and approach the **top human/AI submissions**, normalized
by cross-arch roofline.

```text
GPU:           NVIDIA B200 (leaderboard) vs AMD MI300X/MI355X (this task)
Latency unit:  ms

SOL Bound (theoretical roofline):     0.133004 ms   (SOL-Score=1.0000, speedup=61.76×)
Scoring Baseline (optimized PyTorch): 6.998241 ms   (SOL-Score=0.5146, speedup=1.07×)
Reference Implementation (plain PyTorch): 15.235633 ms   (SOL-Score=0.3267, speedup=0.50×)

Top public submissions:
  - **Rank 1** (`Recursive`): 0.664562 ms, SOL-Score=0.9322, speedup=11.36× vs reference
  - **Rank 2** (`Yatimai`): 0.873276 ms, SOL-Score=0.9084
  - **Rank 3** (`doubleAI`): 1.048443 ms, SOL-Score=0.8888
```

**Cross-arch note:** B200 timings are not directly comparable to MI300X/MI355X
(different GPU arch, different roofline). The agent should:
1. Measure the PyTorch reference on MI300X/MI355X (same shapes, same Docker image)
2. Compute the FlyDSL kernel's speedup vs the on-AMD PyTorch reference
3. Compare that speedup ratio to the leaderboard's `avg_speedup` ratios above
4. Target: FlyDSL speedup vs PyTorch-reference-on-AMD >= 0.8 × (Rank-1 speedup vs PyTorch-reference-on-B200)

## Definition of done

- [ ] FlyDSL kernel passes correctness (max_abs_diff < 1e-2 vs PyTorch reference on all tested workload shapes)
- [ ] FlyDSL kernel_ms measured on MI300X (gfx942) AND MI355X (gfx950) across >=5 workload shapes
- [ ] FlyDSL speedup vs PyTorch-reference-on-AMD >= 0.8 × Rank-1 speedup vs PyTorch-reference-on-B200 (=11.36× if top1 exists)
- [ ] `assessment.md` (<=300 words) ranks FlyDSL vs PyTorch eager on each shape, identifies win/tie/lose regimes
- [ ] Patch modifies actual kernel source (not just bench scripts) — bench-only patches are REJECTED

## Non-goals / rejection criteria

- Bench-only patches (no FlyDSL kernel source changes) are rejected
- Patches that set expected_value = measured_value (tautological pass) are rejected
- The metric contract must compare FlyDSL vs PyTorch reference on AMD, not an absolute ms
- "Completed" without a measured speedup ratio vs the on-AMD PyTorch reference is rejected

**Comment by sunxxuns** [comment]
## AMDPilot acceptance target

**The deliverable is a FlyDSL kernel.** A PyTorch reimplementation of this
problem -- with or without `torch.compile` -- is the *baseline*, not a result.
It does not close this issue no matter how closely it matches the reference.

### Implementation contract (hard requirement)

The submitted kernel MUST be written in FlyDSL and go through the FlyDSL
compiler. Concretely, the kernel module has to actually use the DSL, e.g.:

```python
import flydsl.compiler as flyc
from flydsl.expr import arith as ea, gpu, vector as ev
```

and dispatch through `flyc.jit` / `flyc.kernel`. A file placed under
`FlyDSL/kernels/sol_execbench/` that only calls `torch.*` ops does NOT satisfy
this issue, even if it is numerically perfect and even if it is faster.

If you conclude the port is not expressible in FlyDSL today, say so explicitly
and report exactly which DSL capability is missing. That is a legitimate,
useful outcome. Silently substituting a PyTorch implementation is not.

### Baseline definition

The baseline is the sol-execbench PyTorch reference (`reference.py`) running
eager, with `torch.compile` OFF. Measure it yourself in this run, on the same
GPU and the same workload shapes you benchmark your kernel on. Do not carry a
number over from another run, another GPU, or the issue text.

### Performance target

Your FlyDSL kernel must be **at least 10% faster than that measured baseline**
-- `kernel_ms <= 0.90 x baseline_ms` -- on **at least 80% of the tested
workload shapes**.

### Correctness target

`max_abs_diff < 1e-2` versus the PyTorch reference on **all** tested shapes. A
faster kernel that fails this is a failure, not a tradeoff.

### What you must report

Report a per-shape table containing, for every tested shape: the measured
eager baseline ms, your FlyDSL kernel ms, the ratio, and win/loss. Then state
the win rate against the 80% bar. A result without the baseline column next to
the kernel column is not checkable and will not be credited.
